File: WWF/Hands on labs 3.0/Lab01/Completed/Exercise 3/ExpenseWorkflows/ExpenseLocalServices/ExpenseLocalService.cs
//--------------------------------------------------------------------------------
// This file is a "Sample" as from Windows Workflow Foundation
// Hands on Labs RC
//
// Copyright (c) Microsoft Corporation. All rights reserved.
//
// This source code is intended only as a supplement to Microsoft
// Development Tools and/or on-line documentation.  See these other
// materials for detailed information regarding Microsoft code samples.
// 
// THIS CODE AND INFORMATION ARE PROVIDED AS IS WITHOUT WARRANTY OF ANY
// KIND, EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
// IMPLIED WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A
// PARTICULAR PURPOSE.
//--------------------------------------------------------------------------------

using System;
using System.Collections.Generic;
using System.Text;
using System.Workflow.ComponentModel;
using System.Workflow.Runtime;
using System.Workflow.Runtime.Hosting;

namespace ExpenseLocalServices
{
	public class ExpenseLocalService : IExpenseService
	{
		private WorkflowRuntime _wfruntime;
		private Type _ActivatingWorkflow;
		List<ExpenseReport> _ExpenseReports;

        public ExpenseLocalService(WorkflowRuntime workflowRuntime, Type activatingWorkflow)
        {
			// Store references to both the WorkflowServiceContainer and the Activating Workflow type
			// ...in member variables so we can use them later in this Local Service.  
			_wfruntime = workflowRuntime;
			_ActivatingWorkflow = activatingWorkflow;

			// Create a new List for storing the Expense Reports in memory
			_ExpenseReports = new List<ExpenseReport>();

        }

		#region IExpenseService Members

		void IExpenseService.ApproveExpenseReport(ExpenseReport report)
		{
			// Update the status for the specified report
			UpdateReportStatus(report, StatusEnum.Approved);
		}

		void IExpenseService.RequestManagerApproval(ExpenseReport report, string ManagerEmployeeId)
		{
            // Update the status for the specified report 
			UpdateReportStatus(report, StatusEnum.InReview);
		}

		void IExpenseService.RejectExpenseReport(ExpenseReport report)
		{
            // Update the status for the specified report 
			UpdateReportStatus(report, StatusEnum.Rejected);
		}

		public event EventHandler<ExpenseReportSubmittedEventArgs> ExpenseReportSubmitted;
		public event EventHandler<ExpenseReportReviewedEventArgs> ExpenseReportReviewed;
		#endregion

		private void UpdateReportStatus(ExpenseReport report, StatusEnum status)
		{
			// Find the report specified in our local data cache
			ExpenseReport foundReport = this.GetExpenseReport(report.ExpenseReportId);

			// If we can't find the report, then throw an exception
			if (foundReport == null)
			{
				throw new System.ApplicationException("Unable to find the specified Expense Report");
			}

			// Update the status for the found report
			foundReport.Status = status;
		}
		

		public ExpenseReport GetExpenseReport(System.Guid ExpenseReportId)
		{
			foreach (ExpenseReport report in _ExpenseReports)
			{
				if (report.ExpenseReportId.Equals(ExpenseReportId))
				{
					return report;
				}
			}
			return null;
		}

		public void RaiseExpenseReportSubmittedEvent(System.Guid InstanceId, ExpenseReport report)
		{
			// Start the Expense Reporting workflow
			_wfruntime.CreateWorkflow(this._ActivatingWorkflow, new Dictionary<string,object>(), InstanceId).Start();

			// Create the EventArgs for this event
			ExpenseReportSubmittedEventArgs e = new ExpenseReportSubmittedEventArgs(InstanceId);
			e.Report = report;

			// Raise the ExpenseReportSubmitted event
			if (this.ExpenseReportSubmitted != null)
				this.ExpenseReportSubmitted(null, e);

			// Add the Expense Report to the List
			_ExpenseReports.Add(report);
		}


		public void RaiseExpenseReportReviewedEvent(System.Guid InstanceId, ExpenseReport report, ExpenseReportReview review)
		{

			// Determine the new status for the reviewed report
			StatusEnum status = StatusEnum.InReview;

			if (review.Approved)
			{
				status = StatusEnum.Approved;
			}
			else
			{
				status = StatusEnum.Rejected;
			}

			// Update the status of the report 
			this.UpdateReportStatus(report, status);


			// Create the EventArgs for this event
			ExpenseReportReviewedEventArgs e =
				new ExpenseReportReviewedEventArgs(InstanceId, report, review);

			// Raise the ExpenseReportReviewed event
			if (this.ExpenseReportReviewed != null)
				this.ExpenseReportReviewed(null, e);
		}

		public List<ExpenseReport> GetExpenseReportsList()
		{
			// Return the list of ExpenseReports we've been maintaining in memory
			return _ExpenseReports;
		}
	}
}

File: WWF/Hands on labs 3.0/VB/Lab01/Completed/Exercise 4/ExpenseWorkflows/ExpenseLocalServices/ExpenseLocalService.vb
'--------------------------------------------------------------------------------
' This file is a "Sample" as from Windows Workflow Foundation
' Hands on Labs RC
'
' Copyright (c) Microsoft Corporation. All rights reserved.
'
' This source code is intended only as a supplement to Microsoft
' Development Tools and/or on-line documentation.  See these other
' materials for detailed information regarding Microsoft code samples.
' 
' THIS CODE AND INFORMATION ARE PROVIDED AS IS WITHOUT WARRANTY OF ANY
' KIND, EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
' IMPLIED WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A
' PARTICULAR PURPOSE.
'--------------------------------------------------------------------------------

'--------------------------------------------------------------------------------
' This file is a "Sample" as from Windows Workflow Foundation
' Hands on Labs RC
'
' Copyright (c) Microsoft Corporation. All rights reserved.
'
' This source code is intended only as a supplement to Microsoft
' Development Tools and/or on-line documentation.  See these other
' materials for detailed information regarding Microsoft code samples.
' 
' THIS CODE AND INFORMATION ARE PROVIDED AS IS WITHOUT WARRANTY OF ANY
' KIND, EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
' IMPLIED WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A
' PARTICULAR PURPOSE.
'--------------------------------------------------------------------------------

Imports System.Workflow.Activities
Imports System.Workflow.ComponentModel
Imports System.Workflow.Runtime

Public Class ExpenseLocalService
    Implements IExpenseService

    Private _wfruntime As WorkflowRuntime
    Private _ActivatingWorkflow As Type
    Private _ExpenseReports As List(Of ExpenseReport)

    Public Sub New(ByVal workflowRuntime As WorkflowRuntime, ByVal activatingWorkflow As Type)

        ' Store references to both the WorkflowServiceContainer and the Activating Workflow type
        ' ...in member variables so we can use them later in this Local Service.  
        _wfruntime = WorkflowRuntime
        _ActivatingWorkflow = activatingWorkflow

        ' Create a new List for storing the Expense Reports in memory
        _ExpenseReports = New List(Of ExpenseReport)()
    End Sub

#Region "IExpenseService Members"

    Public Sub ApproveExpenseReport(ByVal report As ExpenseReport) Implements IExpenseService.ApproveExpenseReport
        ' Update the status for the specified report
        UpdateReportStatus(report, StatusEnum.Approved)
    End Sub

    Public Event ExpenseReportSubmitted As EventHandler(Of ExpenseReportSubmittedEventArgs) Implements IExpenseService.ExpenseReportSubmitted

    Public Event ExpenseReportReviewed As EventHandler(Of ExpenseReportReviewedEventArgs) Implements IExpenseService.ExpenseReportReviewed

    Public Sub RejectExpenseReport(ByVal report As ExpenseReport) Implements IExpenseService.RejectExpenseReport
        ' Update the status for the specified report
        UpdateReportStatus(report, StatusEnum.Rejected)
    End Sub

    Public Sub RequestManagerApproval(ByVal report As ExpenseReport, ByVal ManagerEmplyeeId As String) Implements IExpenseService.RequestManagerApproval
        ' Update the status for the specified report
        UpdateReportStatus(report, StatusEnum.InReview)
    End Sub

#End Region

    Private Sub UpdateReportStatus(ByVal report As ExpenseReport, ByVal status As StatusEnum)

        ' Find the report specified in our local data cache
        Dim foundReport As ExpenseReport = Me.GetExpenseReport(report.ExpenseReportId)

        ' If we can't find the report, then throw an exception
        If (foundReport Is Nothing) Then
            Throw New System.ApplicationException("Unable to find the specified Expense Report")
        End If

        ' Update the status for the found report
        foundReport.Status = status
    End Sub

    Public Function GetExpenseReport(ByVal ExpenseReportId As System.Guid) As ExpenseReport
        For Each report As ExpenseReport In _ExpenseReports
            If report.ExpenseReportId.Equals(ExpenseReportId) Then
                Return report
            End If
        Next
        Return Nothing
    End Function

    Public Sub RaiseExpenseReportSubmittedEvent(ByVal InstanceId As System.Guid, ByVal report As ExpenseReport)
        ' Start the Expense Reporting workflow
        _wfruntime.CreateWorkflow(Me._ActivatingWorkflow, New Dictionary(Of String, Object), InstanceId).Start()

        ' Create the EventArgs for this event
        Dim e As ExpenseReportSubmittedEventArgs = New ExpenseReportSubmittedEventArgs(InstanceId)
        e.Report = report

        ' Raise the ExpenseReportSubmitted event
        RaiseEvent ExpenseReportSubmitted(Nothing, e)

        ' Add the Expense Report to the List
        _ExpenseReports.Add(report)
    End Sub


    Public Sub RaiseExpenseReportReviewedEvent(ByVal InstanceId As System.Guid, ByVal report As ExpenseReport, ByVal review As ExpenseReportReview)
        ' Determine the new status for the reviewed report
        Dim status As StatusEnum = StatusEnum.InReview

        If review.Approved Then
            status = StatusEnum.Approved
        Else
            status = StatusEnum.Rejected
        End If

        ' Update the status of the report 
        Me.UpdateReportStatus(report, status)

        ' Create the EventArgs for this event
        Dim e As ExpenseReportReviewedEventArgs = New ExpenseReportReviewedEventArgs(InstanceId, report, review)

        ' Raise the ExpenseReportReviewed event
        RaiseEvent ExpenseReportReviewed(Nothing, e)
    End Sub

    Public Function GetExpenseReportsList() As List(Of ExpenseReport)
        ' Return the list of ExpenseReports we've been maintaining in memory
        Return _ExpenseReports
    End Function


End Class
